examples/llcstat.rs

```diff
@@ -0,0 +1,160 @@
+use bcc::core::BPF;
+use bcc::perf::{PerfHardwareConfig, PerfType};
+use bcc::BccError;
+use clap::{App, Arg};
+
+use core::sync::atomic::{AtomicBool, Ordering};
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::{mem, ptr, str, thread, time};
+
+// Summarize cache reference and cache misses
+//
+// Based on https://github.com/iovisor/bcc/blob/master/tools/llcstat.py
+
+const DEFAULT_SAMPLE_PERIOD: u64 = 100; // Events (Aka every 100 events)
+const DEFAULT_DURATION: u64 = 10; // Seconds
+
+#[repr(C)]
+struct key_t {
+    cpu: i32,
+    pid: i32,
+    name: [u8; 16],
+}
+
+fn do_main(runnable: Arc<AtomicBool>) -> Result<(), BccError> {
+    let matches = App::new("cpudist")
+        .arg(
+            Arg::with_name("sample_period")
+                .long("sample_period")
+                .short("c")
+                .help("Sample one in this many number of cache reference / miss events")
+                .takes_value(true),
+        )
+        .arg(
+            Arg::with_name("duration")
+                .long("duration")
+                .short("d")
+                .help("Duration in seconds to run")
+                .takes_value(true),
+        )
+        .get_matches();
+
+    let sample_period: u64 = matches
+        .value_of("sample_period")
+        .map(|v| v.parse().expect("Invalid sample period"))
+        .unwrap_or(DEFAULT_SAMPLE_PERIOD);
+
+    let duration: u64 = matches
+        .value_of("duration")
+        .map(|v| v.parse().expect("Invalid duration"))
+        .unwrap_or(DEFAULT_DURATION);
+
+    let code = include_str!("llcstat.c").to_string();
+    let mut bpf = BPF::new(&code)?;
+    bpf.attach_perf_event(
+        "on_cache_miss",
+        PerfType::Hardware as u32,
```

Comment:
I don't like having to do `as u32` at the call site. Consider making the function take generic types which have a type bound `u32: From<T>`

Reply:
Deferring this to a future refactor. I need to think about the design more carefully